In [1]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import numpy as np

### Data Overview
The California Department of Education has collected Graduation rate and outcome data that we believe will be invaluable for the purpose of investigating our hypothesis. This data is available per cohort over the duration of their high school career. For example, datasets for 2016-2017 captures data related to the high school cohort that started 9th grade in 2013-2014 and were slated to graduate 2016-2017.

For purpose of our analysis, we will explore the Adjusted Cohort GraduationRate (ACGR) and Outcome Data for 2016-2017, 2017-2018, 2018-2019. Per data source, this data contains multiple levels of aggregation and disaggregation at multiple levels reporting categories.

Also, to protect student privacy, data has been suppressed in instances where * is represented in a file - largely because the cohort size is 10 or less. Also, in instances where a race/ethinicity group is suppressed due to aforementioned, all data in that cluster is not reported for other race/ethnic groups.

#### Data Source
This data is publicly available in .txt format at https://www.cde.ca.gov/ds/sd/sd/filesacgr.asp with a tab delimiters. Data for respective years were downloaded to local destination from site using command line below (skipped checking for ceertificate since a secure connection could not be established and it was blocking access to data):
* wget --no-check-certificate https://www3.cde.ca.gov/demo-downloads/acgr/cohort1617.txt 
* wget --no-check-certificate https://www3.cde.ca.gov/demo-downloads/acgr/cohort1718.txt 
* wget --no-check-certificate https://www3.cde.ca.gov/demo-downloads/acgr/cohort1819.txt 

In [2]:
pwd

'/home/jovyan/work'

#### Data Transformation

In [3]:
#Load data as a DataFrame, convert from .txt to .csv and assign to a variable according to their respective years
GR1617 = pd.read_csv("ProjectData/Graduation_Rate1617.txt",delimiter = "\t")
GR1718 = pd.read_csv("ProjectData/Graduation_Rate1718.txt",delimiter = "\t")
GR1819 = pd.read_csv("ProjectData/Graduation_Rate1819.txt",delimiter = "\t")

A quick look at the data shows very few null elements. There is consistency across which columns have null data as __DistrictCode, SchoolCode, DistrictName__ and __SchoolName__ seem to be the only fields with NULL data across the 3 datasets. Thus, the decision to leave as is and concatenate the 3 dataframes ahead of further analysis and cleaning. The underlying assumption is that the the rows missing this data are those aggregated at the State __"T"__ and County **"C"** level as it seems logical they will not have either Disctrict or School code and/or name. If in the course of cleaning/further analysis this changes, we will update assumption.

Our data source is explicit about having surpressed data using __*__. These do not trasnlate to null as python still reads them as having values. Given this is only specific to categorical variable and are not believed to be able to skew the data in anyway (as they are cross reported under another category), we will retain this entries for now.

In [4]:
print(GR1617.info())
GR1617

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 195869 entries, 0 to 195868
Data columns (total 34 columns):
AcademicYear                               195869 non-null object
AggregateLevel                             195869 non-null object
CountyCode                                 195869 non-null int64
DistrictCode                               187926 non-null float64
SchoolCode                                 187926 non-null float64
CountyName                                 195869 non-null object
DistrictName                               187926 non-null object
SchoolName                                 187926 non-null object
CharterSchool                              195869 non-null object
DASS                                       195869 non-null object
ReportingCategory                          195869 non-null object
CohortStudents                             195869 non-null object
Regular HS Diploma Graduates (Count)       195869 non-null object
Regular HS Diploma Graduates (

AcademicYear AggregateLevel  CountyCode  DistrictCode  SchoolCode  \
0           2016-17              C           1           NaN         NaN   
1           2016-17              C           1           NaN         NaN   
2           2016-17              C           1           NaN         NaN   
3           2016-17              C           1           NaN         NaN   
4           2016-17              C           1           NaN         NaN   
5           2016-17              C           1           NaN         NaN   
6           2016-17              C           1           NaN         NaN   
7           2016-17              C           1           NaN         NaN   
8           2016-17              C           1           NaN         NaN   
9           2016-17              C           1           NaN         NaN   
10          2016-17              C           1           NaN         NaN   
11          2016-17              C           1           NaN         NaN   
12          2016-17              C           1           NaN         NaN   
13          2016-17              C           1           NaN         NaN   
14          2016-17              C           1           NaN         NaN   
15          2016-17              C           1           NaN         NaN   
16          2016-17              C           1           NaN         NaN   
17          2016-17              C           1           NaN         NaN   
18          2016-17              C           1           NaN         NaN   
19          2016-17              C           1           NaN         NaN   
20          2016-17              C           1           NaN         NaN   
21          2016-17              C           1           NaN         NaN   
22          2016-17              C           1           NaN         NaN   
23          2016-17              C           1           NaN         NaN   
24          2016-17              C           1           NaN         NaN   
25          2016-17              C           1           NaN         NaN   
26          2016-17              C           1           NaN         NaN   
27          2016-17              C           1           NaN         NaN   
28          2016-17              C           1           NaN         NaN   
29          2016-17              C           1           NaN         NaN   
...             ...            ...         ...           ...         ...   
195839      2016-17              T           0           NaN         NaN   
195840      2016-17              T           0           NaN         NaN   
195841      2016-17              T           0           NaN         NaN   
195842      2016-17              T           0           NaN         NaN   
195843      2016-17              T           0           NaN         NaN   
195844      2016-17              T           0           NaN         NaN   
195845      2016-17              T           0           NaN         NaN   
195846      2016-17              T           0           NaN         NaN   
195847      2016-17              T           0           NaN         NaN   
195848      2016-17              T           0           NaN         NaN   
195849      2016-17              T           0           NaN         NaN   
195850      2016-17              T           0           NaN         NaN   
195851      2016-17              T           0           NaN         NaN   
195852      2016-17              T           0           NaN         NaN   
195853      2016-17              T           0           NaN         NaN   
195854      2016-17              T           0           NaN         NaN   
195855      2016-17              T           0           NaN         NaN   
195856      2016-17              T           0           NaN         NaN   
195857      2016-17              T           0           NaN         NaN   
195858      2016-17              T           0           NaN         NaN   
195859      2016-17              T           0         

In [5]:
# GR1617.describe(include = ['O'])

In [6]:
print(GR1718.info())
GR1718

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 202115 entries, 0 to 202114
Data columns (total 34 columns):
AcademicYear                               202115 non-null object
AggregateLevel                             202115 non-null object
CountyCode                                 202115 non-null int64
DistrictCode                               194221 non-null float64
SchoolCode                                 194221 non-null float64
CountyName                                 202115 non-null object
DistrictName                               194221 non-null object
SchoolName                                 194221 non-null object
CharterSchool                              202115 non-null object
DASS                                       202115 non-null object
ReportingCategory                          202115 non-null object
CohortStudents                             202115 non-null object
Regular HS Diploma Graduates (Count)       202115 non-null object
Regular HS Diploma Graduates (

AcademicYear AggregateLevel  CountyCode  DistrictCode  SchoolCode  \
0           2017-18              C           1           NaN         NaN   
1           2017-18              C           1           NaN         NaN   
2           2017-18              C           1           NaN         NaN   
3           2017-18              C           1           NaN         NaN   
4           2017-18              C           1           NaN         NaN   
5           2017-18              C           1           NaN         NaN   
6           2017-18              C           1           NaN         NaN   
7           2017-18              C           1           NaN         NaN   
8           2017-18              C           1           NaN         NaN   
9           2017-18              C           1           NaN         NaN   
10          2017-18              C           1           NaN         NaN   
11          2017-18              C           1           NaN         NaN   
12          2017-18              C           1           NaN         NaN   
13          2017-18              C           1           NaN         NaN   
14          2017-18              C           1           NaN         NaN   
15          2017-18              C           1           NaN         NaN   
16          2017-18              C           1           NaN         NaN   
17          2017-18              C           1           NaN         NaN   
18          2017-18              C           1           NaN         NaN   
19          2017-18              C           1           NaN         NaN   
20          2017-18              C           1           NaN         NaN   
21          2017-18              C           1           NaN         NaN   
22          2017-18              C           1           NaN         NaN   
23          2017-18              C           1           NaN         NaN   
24          2017-18              C           1           NaN         NaN   
25          2017-18              C           1           NaN         NaN   
26          2017-18              C           1           NaN         NaN   
27          2017-18              C           1           NaN         NaN   
28          2017-18              C           1           NaN         NaN   
29          2017-18              C           1           NaN         NaN   
...             ...            ...         ...           ...         ...   
202085      2017-18              T           0           NaN         NaN   
202086      2017-18              T           0           NaN         NaN   
202087      2017-18              T           0           NaN         NaN   
202088      2017-18              T           0           NaN         NaN   
202089      2017-18              T           0           NaN         NaN   
202090      2017-18              T           0           NaN         NaN   
202091      2017-18              T           0           NaN         NaN   
202092      2017-18              T           0           NaN         NaN   
202093      2017-18              T           0           NaN         NaN   
202094      2017-18              T           0           NaN         NaN   
202095      2017-18              T           0           NaN         NaN   
202096      2017-18              T           0           NaN         NaN   
202097      2017-18              T           0           NaN         NaN   
202098      2017-18              T           0           NaN         NaN   
202099      2017-18              T           0           NaN         NaN   
202100      2017-18              T           0           NaN         NaN   
202101      2017-18              T           0           NaN         NaN   
202102      2017-18              T           0           NaN         NaN   
202103      2017-18              T           0           NaN         NaN   
202104      2017-18              T           0           NaN         NaN   
202105      2017-18              T           0         

In [7]:
print(GR1819.info())
GR1819

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 198022 entries, 0 to 198021
Data columns (total 34 columns):
AcademicYear                               198022 non-null object
AggregateLevel                             198022 non-null object
CountyCode                                 198022 non-null int64
DistrictCode                               190106 non-null float64
SchoolCode                                 190106 non-null float64
CountyName                                 198022 non-null object
DistrictName                               190106 non-null object
SchoolName                                 190106 non-null object
CharterSchool                              198022 non-null object
DASS                                       198022 non-null object
ReportingCategory                          198022 non-null object
CohortStudents                             198022 non-null object
Regular HS Diploma Graduates (Count)       198022 non-null object
Regular HS Diploma Graduates (

AcademicYear AggregateLevel  CountyCode  DistrictCode  SchoolCode  \
0           2018-19              C           1           NaN         NaN   
1           2018-19              C           1           NaN         NaN   
2           2018-19              C           1           NaN         NaN   
3           2018-19              C           1           NaN         NaN   
4           2018-19              C           1           NaN         NaN   
5           2018-19              C           1           NaN         NaN   
6           2018-19              C           1           NaN         NaN   
7           2018-19              C           1           NaN         NaN   
8           2018-19              C           1           NaN         NaN   
9           2018-19              C           1           NaN         NaN   
10          2018-19              C           1           NaN         NaN   
11          2018-19              C           1           NaN         NaN   
12          2018-19              C           1           NaN         NaN   
13          2018-19              C           1           NaN         NaN   
14          2018-19              C           1           NaN         NaN   
15          2018-19              C           1           NaN         NaN   
16          2018-19              C           1           NaN         NaN   
17          2018-19              C           1           NaN         NaN   
18          2018-19              C           1           NaN         NaN   
19          2018-19              C           1           NaN         NaN   
20          2018-19              C           1           NaN         NaN   
21          2018-19              C           1           NaN         NaN   
22          2018-19              C           1           NaN         NaN   
23          2018-19              C           1           NaN         NaN   
24          2018-19              C           1           NaN         NaN   
25          2018-19              C           1           NaN         NaN   
26          2018-19              C           1           NaN         NaN   
27          2018-19              C           1           NaN         NaN   
28          2018-19              C           1           NaN         NaN   
29          2018-19              C           1           NaN         NaN   
...             ...            ...         ...           ...         ...   
197992      2018-19              T           0           NaN         NaN   
197993      2018-19              T           0           NaN         NaN   
197994      2018-19              T           0           NaN         NaN   
197995      2018-19              T           0           NaN         NaN   
197996      2018-19              T           0           NaN         NaN   
197997      2018-19              T           0           NaN         NaN   
197998      2018-19              T           0           NaN         NaN   
197999      2018-19              T           0           NaN         NaN   
198000      2018-19              T           0           NaN         NaN   
198001      2018-19              T           0           NaN         NaN   
198002      2018-19              T           0           NaN         NaN   
198003      2018-19              T           0           NaN         NaN   
198004      2018-19              T           0           NaN         NaN   
198005      2018-19              T           0           NaN         NaN   
198006      2018-19              T           0           NaN         NaN   
198007      2018-19              T           0           NaN         NaN   
198008      2018-19              T           0           NaN         NaN   
198009      2018-19              T           0           NaN         NaN   
198010      2018-19              T           0           NaN         NaN   
198011      2018-19              T           0           NaN         NaN   
198012      2018-19              T           0         

In [8]:
years = [GR1617, GR1718, GR1819]
combinedGR = pd.concat(years, ignore_index=True)
#ignore_index resets the numbering of the rows
pd.set_option('max_columns', None)
#Quick view at head and tail
combinedGR.head(20).append(combinedGR.tail(20))

AcademicYear AggregateLevel  CountyCode  DistrictCode  SchoolCode  \
0           2016-17              C           1           NaN         NaN   
1           2016-17              C           1           NaN         NaN   
2           2016-17              C           1           NaN         NaN   
3           2016-17              C           1           NaN         NaN   
4           2016-17              C           1           NaN         NaN   
5           2016-17              C           1           NaN         NaN   
6           2016-17              C           1           NaN         NaN   
7           2016-17              C           1           NaN         NaN   
8           2016-17              C           1           NaN         NaN   
9           2016-17              C           1           NaN         NaN   
10          2016-17              C           1           NaN         NaN   
11          2016-17              C           1           NaN         NaN   
12          2016-17              C           1           NaN         NaN   
13          2016-17              C           1           NaN         NaN   
14          2016-17              C           1           NaN         NaN   
15          2016-17              C           1           NaN         NaN   
16          2016-17              C           1           NaN         NaN   
17          2016-17              C           1           NaN         NaN   
18          2016-17              C           1           NaN         NaN   
19          2016-17              C           1           NaN         NaN   
595986      2018-19              T           0           NaN         NaN   
595987      2018-19              T           0           NaN         NaN   
595988      2018-19              T           0           NaN         NaN   
595989      2018-19              T           0           NaN         NaN   
595990      2018-19              T           0           NaN         NaN   
595991      2018-19              T           0           NaN         NaN   
595992      2018-19              T           0           NaN         NaN   
595993      2018-19              T           0           NaN         NaN   
595994      2018-19              T           0           NaN         NaN   
595995      2018-19              T           0           NaN         NaN   
595996      2018-19              T           0           NaN         NaN   
595997      2018-19              T           0           NaN         NaN   
595998      2018-19              T           0           NaN         NaN   
595999      2018-19              T           0           NaN         NaN   
596000      2018-19              T           0           NaN         NaN   
596001      2018-19              T           0           NaN         NaN   
596002      2018-19              T           0           NaN         NaN   
596003      2018-19              T           0           NaN         NaN   
596004      2018-19              T           0           NaN         NaN   
596005      2018-19              T           0           NaN         NaN   

       CountyName DistrictName SchoolName CharterSchool DASS  \
0         Alameda          NaN        NaN           All  All   
1         Alameda          NaN        NaN           All  All   
2         Alameda          NaN        NaN           All  All   
3         Alameda          NaN        NaN           All  All   
4         Alameda          NaN        NaN           All  All   
5         Alameda          NaN        NaN           All  All   
6         Alameda          NaN        NaN           All  All   
7         Alameda          NaN        NaN           All  All   
8         Alameda          NaN        NaN           All  All   
9         Alameda          NaN        NaN           All  All   
10        Alameda          NaN        NaN           All  All   
11        Alameda          NaN        NaN           All  All   
12        Alameda          NaN        NaN           All  A

In [9]:
combinedGR.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 596006 entries, 0 to 596005
Data columns (total 34 columns):
AcademicYear                               596006 non-null object
AggregateLevel                             596006 non-null object
CountyCode                                 596006 non-null int64
DistrictCode                               572253 non-null float64
SchoolCode                                 572253 non-null float64
CountyName                                 596006 non-null object
DistrictName                               572253 non-null object
SchoolName                                 572253 non-null object
CharterSchool                              596006 non-null object
DASS                                       596006 non-null object
ReportingCategory                          596006 non-null object
CohortStudents                             596006 non-null object
Regular HS Diploma Graduates (Count)       596006 non-null object
Regular HS Diploma Graduates (

********
#### Data Description

Our combined DF has **596,005** rows and **34** columns. Below is a definition of the columns stating with the categorical variables that are basisi of aggregation and disaggregation/subset for this dataset and must be carefully selected during analysis in order to obtain an unduplicated count:

###### Categorical variables
1. **AggregateLevel**: This data is aggregated at 4 categorical levels. <font color='red'> _In order to obtain and unduplicated count, one of these levels must be chosen in analyzing this data_ </font>:
    * __T__ = State
    * __C__ = County
    * __D__ = District
    * __S__ = School
    
* __CharterSchool__: This column is indicative of whether data in that category encompasses for all schools, only charter schools or non-charter schools. <font color='red'>_In order to obtain and unduplicated count at the selected level, one of these values must be chosen in analyzing this dataset_ </font>:
     - __All__ = Includes data for __all schools__ regardless of charter status. For AggregateLevel=S, all schools are included. For Aggregate Levels T, C, and D, data for allschools are included in the aggregate.
     - __Yes__ = Includes __only data for charter schools__. For Aggregate Level=S, __only charter schools__ are included. For Aggregate Levels T, C, and D, __only data for charter schools__ are included in the aggregate.
     - __No__ = Includes __only data for non-charter schools__. For Aggregate Level=S, __only non-charter schools__ are included. For Aggregate Levels T, C, and D, __only datafor non-charter schools__ are included in the aggregate.
     
*  __DASS__: An indicator of whether data for all schools, only schools participating in theDashboard Alternative School Status Program (DASS), or only non-DASS schools are included. <font color='red'>_In order to obtain and unduplicated count at the selected level, one of these values must be chosen in analyzing this dataset_ </font>: 
    - __All__ = Includes data for __all schools regardless of DASS status__. For AggregateLevel=S, all schools are included. For Aggregate Levels T, C, and D, data for allschools are included in the aggregate.
    - __Yes__ = Includes __only data for schools participating in DASS__. For AggregateLevel=S, __only DASS schools__ are included. For Aggregate Levels T, C, and D, __only data for DASS schools__ are included in the aggregate.
    - __No__ = Includes __only data for non-DASS schools__. For Aggregate Level=S, __only data for non-DASS schools__ are included. For Aggregate Levels T, C, and D, __only data for non-DASS schools__ are included in the aggregate.
    
* __ReportingCategory__: This category helps disaggregate this data into select Race, Ethnicity, gender or program subgroup spanning 18 categories. _If not interested in exploring any of the underlisted categories, Reporting Category should be set to TA_:
    - __RB__ = African American
    - __RI__ = American Indian or Alaska Native
    - __RA__ = Asian
    - __RF__ = Filipino
    - __RH__ = Hispanic or Latino
    - __RD__ = Not Reported
    - __RP__ = Pacific Islander
    - __RT__ = Two or More Races
    - __RW__ = White
    - __GM__ = Male
    - __GF__ = Female
    - __SE__ = English Learners
    - __SD__ = Students with Disabilities
    - __SS__ = Socioeconomically Disadvantaged
    - __SM__ = Migrant
    - __SF__ = Foster
    - __SH__ = Homeless
    - __TA__ = Total
********
##### Other columns    
5. **CountyCode**: A unique two-digit code corresponding to the county.   

6. **DistrictCode**: A unique five-digit code corresponding to the district.

* __SchoolCode__: A unique seven-digit code corresponding to the school.

* __CountyName__: County name.

* __DistrictName__: Ditrict or Administrative authority name.

* __SchoolName__: School name.

* __AcademicYear__: expected year of on-time, four-year graduation from high school

* __CohortStudents__: The number of students who enter grade 9 for the first time, plus any students who transfer in later during grade 9 or the next three years. This does not included any students from the cohort who transfers out, emigrates from the country, transfers to a juvenile facility or prison, or dies during that same period.                           
* __Regular HS Diploma Graduates (Count)__ : Total number of cohort students who received the standard high school diploma from a school aligned with State education standards. <font color="green">_This is a reference columns for a lot of the other columns in  this dataset_ </font>
* __Regular HS Diploma Graduates (Rate)__: The percentage of cohort students who received the standard high school diploma.
* __Met UC/CSU Grad Req's (Count)__: Total number of cohort graduates who met all a-g requirements for admission into a UC or CSU school.
* __Met UC/CSU Grad Req's (Rate)__: The percentage of cohort graduates who met admission requirements for a UC or CSU school.
* __Seal of Biliteracy (Count)__: Total number of cohort graduates who attained a high level of proficiency in one or more languages in addition to English and earned the State Seal of Biliteracy.
* __Seal of Biliteracy (Rate)__: The percentage of cohort graduates who earned the State Seal of Biliteracy.
* __Golden State Seal Merit Diploma (Count)__: Total number of cohort graduates who earned the Golden State Seal Merit Diploma by earning a high school diploma and demonstrating mastery in at leastsix subject matter areas, including mathematics, English language arts, science,U.S. history, and other areas. 
* __Golden State Seal Merit Diploma (Rate)__: The percentage of cohort graduates who earned the Golden State Seal Merit Diploma. 
* __CHSPE Completer (Count)__: Total number of cohort students who withdrew from regular high school without receiving a regular high school diploma after passing the CA High School Proficiency Exam. 
* __CHSPE Completer (Rate)__: The percentage of cohort students who took and passed the CHSPE.
* __Adult Ed. HS Diploma (Count)__: Total number of cohort students who withdrew from regular high school without receiving a regular high school diploma, and later received documentation for completing an adult education program.
* __Adult Ed. HS Diploma (Rate)__: The percentage of cohort students who received documentation for completing an adult education program.
* __SPED Certificate (Count)__: Total number of cohort students with exceptional needs (having an IEP) who received a certificate or document of education achievement or completion.
* __SPED Certificate (Rate)__: The percentage of cohort student with exceptional needs who received a certificate or document of education achievement or completion.
* __GED Completer (Count)__: Total number of cohort students who withdrew from regular high school without receiving a regular high school diploma and has received a High School Equivalency Certificate by passing the GED exam, the TASC exam, and/or the HiSET.
* __GED Completer (Rate)__: The percentage of cohort students who received a High School Equivalency Certificate. 
* __Other Transfer (Count)__: Total number of cohort students who withdrew from regular high school without receiving a regular high school diploma and transfer to an adult education programor to community college during the cohort period. 
* __Other Transfer (Rate)__: The percentage of cohort students who transfer to an adult education program orto community college during the cohort period.
* __Dropout (Count)__: Total number of cohort students who do not graduate with a regular high school diploma, do not complete high school, and are not still enrolled as a "fifth yearsenior".
* __Dropout (Rate)__: The percentage of cohort students who do not graduate with a regular high school diploma, do not complete high school, and are not still enrolled as a "fifth yearsenior".
* __Still Enrolled (Count)__: Total number of cohort students who did not graduate high school by the end of their cohort outcome period who: are re-enrolled as a "fifth year senior" at the endof their cohort period, or who completed high school no later than mid-Septemberof the subsequent academic year.
* __Still Enrolled (Rate)__: The percentage of cohort students who did not graduate high school at the end of their cohort outcome period who either re-enrolled or completed high school nolater than mid-September of the next academic year.   

****************
*The following additional information is also shared on by CDE as added note about using the ACGR and Outcome Data Files:*

This file contains different subsets of data based on the values of different variables. If these sub-setting variables are not used properly, it is easy to end up with incorrect or duplicative data. Please refer to the following examples below as a guide for properly using these data:
* Get school-level data totals with one record per school with school totals (without studentgroup totals)
    - a. Aggregate Level = S (School)
    - b. Charter School (Y/N) = All
    - c. DASS (Y/N) = All
    - d. Reporting Category = TA
* Get school-level data totals by student group (Reporting Category) for all non-charter schools
    - a. Aggregate Level = S (School)
    - b. Charter School (Y/N) = N
    - c. DASS (Y/N) = All
    - d. Reporting Category Not equal to TA   
* Get school-level data totals by student group (Reporting Category) for all DASS schools
    - a. Aggregate Level = S (School)
    - b. Charter School (Y/N) = All
    - c. DASS (Y/N) = Y
    - d. Reporting Category Not equal to TA
* Get district-level data totals by student group (Reporting Category) for non-charter schools
    - a. Aggregate Level = D (District)
    - b. Charter School (Y/N) = N
    - c. DASS (Y/N) = All
    - d. Reporting Category Not equal to TA
* Get statewide-level data totals for charter schools that are DASS schools
    - a. Aggregate Level = T (State)
    - b. Charter School (Y/N) = Y
    - c. DASS (Y/N) = Y
    - d. Reporting Category = TA
    
*******************

### Data Cleaning
#### <font color = 'Green'> Drop Columns </font>
As seen from the data columns description above, there are some columns (Rate) that are calculated fields derived from the count columns. Given these fields do not yet serve any relevance for the purpose of our analysis and can be easily recaculated/derived, we will deem these irrelevant observations for now and drop the columns.

This action leaves us with 23 columns that are considered important for our immediate purpose as opposed to the 34 columns we started with.

In [10]:
#Drop columns using a combination of .drop and .iloc and assign remaining columns to a new DataFrame
NoRatesCombinedGR = combinedGR.drop(combinedGR.iloc[:, [13, 15, 17, 19, 21, 23, 25, 27, 29, 31, 33]], axis=1)

NoRatesCombinedGR.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 596006 entries, 0 to 596005
Data columns (total 23 columns):
AcademicYear                               596006 non-null object
AggregateLevel                             596006 non-null object
CountyCode                                 596006 non-null int64
DistrictCode                               572253 non-null float64
SchoolCode                                 572253 non-null float64
CountyName                                 596006 non-null object
DistrictName                               572253 non-null object
SchoolName                                 572253 non-null object
CharterSchool                              596006 non-null object
DASS                                       596006 non-null object
ReportingCategory                          596006 non-null object
CohortStudents                             596006 non-null object
Regular HS Diploma Graduates (Count)       596006 non-null object
Met UC/CSU Grad Req's (Count) 

In [11]:
NoRatesCombinedGR.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 596006 entries, 0 to 596005
Data columns (total 23 columns):
AcademicYear                               596006 non-null object
AggregateLevel                             596006 non-null object
CountyCode                                 596006 non-null int64
DistrictCode                               572253 non-null float64
SchoolCode                                 572253 non-null float64
CountyName                                 596006 non-null object
DistrictName                               572253 non-null object
SchoolName                                 572253 non-null object
CharterSchool                              596006 non-null object
DASS                                       596006 non-null object
ReportingCategory                          596006 non-null object
CohortStudents                             596006 non-null object
Regular HS Diploma Graduates (Count)       596006 non-null object
Met UC/CSU Grad Req's (Count) 

#### <font color = 'Green'> Fix Structural issues and Data Types </font>
Examine the column labels and values for inconsistency that may hamper our analysis or indvertently skew outcome of our analysis - starting with the categorical variables.

In [12]:
NoRatesCombinedGR.DASS.unique()

array(['All', 'No ', 'Yes'], dtype=object)

In [13]:
#As seen above, there is a trailing space that needs to be stripped from "No  " to ensure consistency and avoid the risk of not returning any values when called without the trailing spaces
#Strip space from DASS using .str.rstrip
NoRatesCombinedGR['DASS'] = NoRatesCombinedGR['DASS'].str.rstrip()
NoRatesCombinedGR.DASS.unique()

array(['All', 'No', 'Yes'], dtype=object)

In [14]:
NoRatesCombinedGR.CharterSchool.unique()

array(['All', 'No ', 'Yes'], dtype=object)

In [15]:
#Strip space from CharterSchool using .str.rstrip
NoRatesCombinedGR['CharterSchool'] = NoRatesCombinedGR['CharterSchool'].str.rstrip()
NoRatesCombinedGR.CharterSchool.unique()

array(['All', 'No', 'Yes'], dtype=object)

In [16]:
NoRatesCombinedGR.AggregateLevel.unique()
#There is consistency in the formart of the AggregateLevel

array(['C', 'D', 'S', 'T'], dtype=object)

In [17]:
NoRatesCombinedGR.ReportingCategory.unique()
#The output below shows there is an additonal category "GX" not expected in our data so will proceed with a closer look

array(['GF', 'GM', 'RA', 'RB', 'RD', 'RF', 'RH', 'RI', 'RP', 'RT', 'RW',
       'SD', 'SE', 'SF', 'SH', 'SM', 'SS', 'TA', 'GX'], dtype=object)

In [18]:
#Isolate the rows that have entires of "GX" as ReportingCategory and add other parameters to ensure they are unduplicated
# GX = NoRatesCombinedGR.loc[(NoRatesCombinedGR['ReportingCategory']== "GX") & (NoRatesCombinedGR['AggregateLevel'] == "S") 
#                       & (NoRatesCombinedGR['CharterSchool'] == "All") & (NoRatesCombinedGR['DASS'] == "All")]

# GX

#Interestng to highlight that this cell and the next produce the same data albeit different indices. 
#Adding in comment to highlight how it can be possible to have a duplicated count.

In [19]:
#Isolate the rows that have entires of "GX" as ReportingCategory and add other parameters to ensure they are unduplicated
GX = NoRatesCombinedGR.loc[(NoRatesCombinedGR['ReportingCategory']== "GX") & (NoRatesCombinedGR['AggregateLevel'] == "S") 
                      & (NoRatesCombinedGR['CharterSchool'] != "All") & (NoRatesCombinedGR['DASS'] != "All")]

GX

AcademicYear AggregateLevel  CountyCode  DistrictCode  SchoolCode  \
65416       2016-17              S           7       61754.0    730044.0   
483271      2018-19              S          19       64287.0   1996479.0   
509052      2018-19              S          19       75309.0    136648.0   

          CountyName              DistrictName  \
65416   Contra Costa        Mt. Diablo Unified   
483271   Los Angeles      Baldwin Park Unified   
509052   Los Angeles  Acton-Agua Dulce Unified   

                                       SchoolName CharterSchool DASS  \
65416                              Northgate High            No   No   
483271  Opportunities for Learning - Baldwin Park           Yes  Yes   
509052                    Options for Youth-Acton           Yes  Yes   

       ReportingCategory CohortStudents Regular HS Diploma Graduates (Count)  \
65416                 GX              *                                    *   
483271                GX              *                                    *   
509052                GX              *                                    *   

       Met UC/CSU Grad Req's (Count) Seal of Biliteracy (Count)  \
65416                              *                          *   
483271                             *                          *   
509052                             *                          *   

       Golden State Seal Merit Diploma (Count) CHSPE Completer (Count)  \
65416                                        *                       *   
483271                                       *                       *   
509052                                       *                       *   

       Adult Ed. HS Diploma (Count) SPED Certificate (Count)  \
65416                             *                        *   
483271                            *                        *   
509052                            *                        *   

       GED Completer (Count) Other Transfer (Count) Dropout (Count)  \
65416                      *                      *               *   
483271                     *                      *               *   
509052                     *                      *               *   

       Still Enrolled (Count)  
65416                       *  
483271                      *  
509052                      *

From the above, we see that we have 3 entires of "GX" category across 3 schools but all their data is supressed. At this point, it can be determined these rows will add no additonal information in the course of our analysis and dropping them will not lead to loss of any valuable input since same data should be represented in the "TA" category for those schools across same years. Thus, the decision to drop.

In [20]:
#Drop the index for the GX caterories identified above and renumber the rows
NoRatesCombinedGR = NoRatesCombinedGR.drop([NoRatesCombinedGR.index[65416], NoRatesCombinedGR.index[483271], NoRatesCombinedGR.index[509052]]).reset_index(drop=True)

In [21]:
NoRatesCombinedGR['CountyCode'].unique()

array([ 1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16, 17,
       18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34,
       35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51,
       52, 53, 54, 55, 56, 57, 58,  0])

In [22]:
#Ensuring all CountyCode are 2 digit as expected
NoRatesCombinedGR['CountyCode'] = (NoRatesCombinedGR['CountyCode']).apply(lambda x: '{0:0>2}'.format(x))
print(NoRatesCombinedGR['CountyCode'])

0         01
1         01
2         01
3         01
4         01
5         01
6         01
7         01
8         01
9         01
10        01
11        01
12        01
13        01
14        01
15        01
16        01
17        01
18        01
19        01
20        01
21        01
22        01
23        01
24        01
25        01
26        01
27        01
28        01
29        01
          ..
595973    00
595974    00
595975    00
595976    00
595977    00
595978    00
595979    00
595980    00
595981    00
595982    00
595983    00
595984    00
595985    00
595986    00
595987    00
595988    00
595989    00
595990    00
595991    00
595992    00
595993    00
595994    00
595995    00
595996    00
595997    00
595998    00
595999    00
596000    00
596001    00
596002    00
Name: CountyCode, Length: 596003, dtype: object


In [23]:
NoRatesCombinedGR['DistrictCode'].describe()

count    572250.000000
mean      62827.538770
std       16989.007069
min       10017.000000
25%       64279.000000
50%       67082.000000
75%       70524.000000
max       77222.000000
Name: DistrictCode, dtype: float64

In [24]:
NoRatesCombinedGR['DistrictCode']

0        NaN
1        NaN
2        NaN
3        NaN
4        NaN
5        NaN
6        NaN
7        NaN
8        NaN
9        NaN
10       NaN
11       NaN
12       NaN
13       NaN
14       NaN
15       NaN
16       NaN
17       NaN
18       NaN
19       NaN
20       NaN
21       NaN
22       NaN
23       NaN
24       NaN
25       NaN
26       NaN
27       NaN
28       NaN
29       NaN
          ..
595973   NaN
595974   NaN
595975   NaN
595976   NaN
595977   NaN
595978   NaN
595979   NaN
595980   NaN
595981   NaN
595982   NaN
595983   NaN
595984   NaN
595985   NaN
595986   NaN
595987   NaN
595988   NaN
595989   NaN
595990   NaN
595991   NaN
595992   NaN
595993   NaN
595994   NaN
595995   NaN
595996   NaN
595997   NaN
595998   NaN
595999   NaN
596000   NaN
596001   NaN
596002   NaN
Name: DistrictCode, Length: 596003, dtype: float64

In [25]:
#Need to normalize the DistrictCode and SchoolCode as well

In [26]:
NoRatesCombinedGR['SchoolCode'].describe()

count    5.722500e+05
mean     1.566504e+06
std      1.867990e+06
min      0.000000e+00
25%      0.000000e+00
50%      1.360510e+05
75%      3.134657e+06
max      9.010745e+06
Name: SchoolCode, dtype: float64

###### <font color='green'> Revisiting supressed data (__*__) </font>

Attempts to convert the other columns to numeric to enable calculation on them was met with an error because of the wildcard used to surpress some data. Hence the following:

In [27]:
#A Selection of * on CohortStudents at aggregate level "S"
NoRatesCombinedGR.loc[(NoRatesCombinedGR['AggregateLevel'] == "S")
                                        & (NoRatesCombinedGR['CohortStudents'] == "*") & (NoRatesCombinedGR['CharterSchool'] != "All") 
                                          & (NoRatesCombinedGR['DASS'] != "All")]

AcademicYear AggregateLevel CountyCode  DistrictCode  SchoolCode  \
55376       2016-17              S         01       10017.0    112607.0   
55378       2016-17              S         01       10017.0    112607.0   
55380       2016-17              S         01       10017.0    112607.0   
55381       2016-17              S         01       10017.0    112607.0   
55382       2016-17              S         01       10017.0    112607.0   
55384       2016-17              S         01       10017.0    112607.0   
55435       2016-17              S         01       10017.0    130401.0   
55436       2016-17              S         01       10017.0    130401.0   
55438       2016-17              S         01       10017.0    130401.0   
55439       2016-17              S         01       10017.0    130401.0   
55444       2016-17              S         01       10017.0    130401.0   
55494       2016-17              S         01       10017.0    130419.0   
55496       2016-17              S         01       10017.0    130419.0   
55498       2016-17              S         01       10017.0    130419.0   
55499       2016-17              S         01       10017.0    130419.0   
55500       2016-17              S         01       10017.0    130419.0   
55543       2016-17              S         01       31609.0    131755.0   
55544       2016-17              S         01       31609.0    131755.0   
55545       2016-17              S         01       31609.0    131755.0   
55546       2016-17              S         01       31609.0    131755.0   
55547       2016-17              S         01       31609.0    131755.0   
55548       2016-17              S         01       31609.0    131755.0   
55549       2016-17              S         01       31609.0    131755.0   
55550       2016-17              S         01       31609.0    131755.0   
55552       2016-17              S         01       31609.0    131755.0   
55602       2016-17              S         01       31617.0    131763.0   
55603       2016-17              S         01       31617.0    131763.0   
55604       2016-17              S         01       31617.0    131763.0   
55606       2016-17              S         01       31617.0    131763.0   
55607       2016-17              S         01       31617.0    131763.0   
...             ...            ...        ...           ...         ...   
595629      2018-19              S         58       72736.0   5830096.0   
595630      2018-19              S         58       72736.0   5830096.0   
595632      2018-19              S         58       72736.0   5830096.0   
595633      2018-19              S         58       72736.0   5830096.0   
595638      2018-19              S         58       72736.0   5830096.0   
595676      2018-19              S         58       72736.0   5830138.0   
595677      2018-19              S         58       72736.0   5830138.0   
595679      2018-19              S         58       72736.0   5830138.0   
595681      2018-19              S         58       72736.0   5830138.0   
595682      2018-19              S         58       72736.0   5830138.0   
595739      2018-19              S         58       72736.0   5835202.0   
595740      2018-19              S         58       72736.0   5835202.0   
595743      2018-19              S         58       72736.0   5835202.0   
595744      2018-19              S         58       72736.0   5835202.0   
595748      2018-19              S         58       72736.0   5835202.0   
595749      2018-19              S         58       72736.0   5835202.0   
595750      2018-19              S         58       72736.0   5835202.0   
595765      2018-19              S         58       72769.0         1.0   
595766      2018-19              S         58       72769.0         1.0   
595767      2018-19              S         58       72769.0         1.0   
595768      2018-19              S         58       72769.0         1.0   
595822      2018-19             

With the aggregate level set to __"S"__ and all other parameters above considered, 52,846 unduplicated rows have supressed data as shown above. Of these rows, only 1,548 fall within the subsets of ReportingCategory = "TA". This means that 97% of this surpressed data has been accounted for in the TA category. Per the note below (from the data source), it can be inferred that the 1,548 surpressed rows is due to the cohort student size being 10 or less. 

__Note__: <font color = 'red'> To protect student privacy, data are suppressed (*) on the data file if the cell size within aselected student population (cohort students) is 10 or less. Additionally, the "Not Reported"race/ethnicity is suppressed, regardless of actual cell size, if the student population for one or moreother race/ethnicity groups is suppressed. </font>

Given that the CohortStudent count is a column that might eventually become relevant in evaluating funding, we will save this to a separate DataFrame and drop the rows for now. The other 51,298 rows will be deleted with the assumption that they have been accounted for in the TA category for their respective schools.

At other levels of aggregation the suppressed value __*__ is intentionally replaced with NaN to allow for conversion of DF to numeric datatype and allow for subsequent calculations on those columns.

In [28]:
SuppressedntTA = NoRatesCombinedGR.loc[(NoRatesCombinedGR['AggregateLevel'] == "S")
                                        & (NoRatesCombinedGR['CohortStudents'] == "*") & (NoRatesCombinedGR['CharterSchool'] != "All") 
                                          & (NoRatesCombinedGR['DASS'] != "All") & (NoRatesCombinedGR['ReportingCategory'] != "TA")]

SuppressedntTA_Index = SuppressedntTA.index
SuppressedntTA_Index

Int64Index([ 55376,  55378,  55380,  55381,  55382,  55384,  55435,  55436,
             55438,  55439,
            ...
            595767, 595822, 595823, 595824, 595825, 595827, 595828, 595832,
            595833, 595834],
           dtype='int64', length=51298)

In [29]:
#This is our dataset minus the surpressed data in other Reporting Caterories except 'TA" which is believed to have been accounted for in the ReportingCategory "TA"
GRminusSur = NoRatesCombinedGR.drop(SuppressedntTA_Index).reset_index(drop=True)
GRminusSur

AcademicYear AggregateLevel CountyCode  DistrictCode  SchoolCode  \
0           2016-17              C         01           NaN         NaN   
1           2016-17              C         01           NaN         NaN   
2           2016-17              C         01           NaN         NaN   
3           2016-17              C         01           NaN         NaN   
4           2016-17              C         01           NaN         NaN   
5           2016-17              C         01           NaN         NaN   
6           2016-17              C         01           NaN         NaN   
7           2016-17              C         01           NaN         NaN   
8           2016-17              C         01           NaN         NaN   
9           2016-17              C         01           NaN         NaN   
10          2016-17              C         01           NaN         NaN   
11          2016-17              C         01           NaN         NaN   
12          2016-17              C         01           NaN         NaN   
13          2016-17              C         01           NaN         NaN   
14          2016-17              C         01           NaN         NaN   
15          2016-17              C         01           NaN         NaN   
16          2016-17              C         01           NaN         NaN   
17          2016-17              C         01           NaN         NaN   
18          2016-17              C         01           NaN         NaN   
19          2016-17              C         01           NaN         NaN   
20          2016-17              C         01           NaN         NaN   
21          2016-17              C         01           NaN         NaN   
22          2016-17              C         01           NaN         NaN   
23          2016-17              C         01           NaN         NaN   
24          2016-17              C         01           NaN         NaN   
25          2016-17              C         01           NaN         NaN   
26          2016-17              C         01           NaN         NaN   
27          2016-17              C         01           NaN         NaN   
28          2016-17              C         01           NaN         NaN   
29          2016-17              C         01           NaN         NaN   
...             ...            ...        ...           ...         ...   
544675      2018-19              T         00           NaN         NaN   
544676      2018-19              T         00           NaN         NaN   
544677      2018-19              T         00           NaN         NaN   
544678      2018-19              T         00           NaN         NaN   
544679      2018-19              T         00           NaN         NaN   
544680      2018-19              T         00           NaN         NaN   
544681      2018-19              T         00           NaN         NaN   
544682      2018-19              T         00           NaN         NaN   
544683      2018-19              T         00           NaN         NaN   
544684      2018-19              T         00           NaN         NaN   
544685      2018-19              T         00           NaN         NaN   
544686      2018-19              T         00           NaN         NaN   
544687      2018-19              T         00           NaN         NaN   
544688      2018-19              T         00           NaN         NaN   
544689      2018-19              T         00           NaN         NaN   
544690      2018-19              T         00           NaN         NaN   
544691      2018-19              T         00           NaN         NaN   
544692      2018-19              T         00           NaN         NaN   
544693      2018-19              T         00           NaN         NaN   
544694      2018-19              T         00           NaN         NaN   
544695      2018-19              T         00           NaN         NaN   
544696      2018-19             

In [30]:
SuppressedTA = GRminusSur.loc[(GRminusSur['CohortStudents'] == "*") & (GRminusSur['ReportingCategory'] == "TA") 
                         & (GRminusSur['AggregateLevel'] == "S") & (GRminusSur['CharterSchool'] != "All")
                            & (GRminusSur['DASS'] != "All")]
SuppressedTA

AcademicYear AggregateLevel CountyCode  DistrictCode  SchoolCode  \
55606       2016-17              S         01       61119.0         1.0   
55962       2016-17              S         01       61127.0         1.0   
55996       2016-17              S         01       61127.0    130294.0   
56071       2016-17              S         01       61143.0         1.0   
56242       2016-17              S         01       61150.0         1.0   
56325       2016-17              S         01       61150.0    133876.0   
56452       2016-17              S         01       61176.0         1.0   
56471       2016-17              S         01       61176.0    127233.0   
56933       2016-17              S         01       61192.0         1.0   
57345       2016-17              S         01       61200.0         1.0   
57806       2016-17              S         01       61242.0         1.0   
57893       2016-17              S         01       61242.0    110478.0   
58154       2016-17              S         01       61259.0    106542.0   
58668       2016-17              S         01       61259.0    121376.0   
59037       2016-17              S         01       61259.0    132555.0   
59428       2016-17              S         01       61275.0         1.0   
59537       2016-17              S         01       61291.0         1.0   
59666       2016-17              S         01       61309.0         1.0   
60099       2016-17              S         01       75101.0         1.0   
60279       2016-17              S         02       10025.0    230011.0   
60295       2016-17              S         03       10033.0    330035.0   
60323       2016-17              S         03       10033.0    330068.0   
60568       2016-17              S         04       10041.0    129817.0   
60727       2016-17              S         04       10041.0   6069256.0   
60752       2016-17              S         04       61408.0    430207.0   
60822       2016-17              S         04       61424.0    111039.0   
60976       2016-17              S         04       61424.0    430215.0   
61284       2016-17              S         04       61515.0         1.0   
61528       2016-17              S         04       61531.0    430108.0   
61860       2016-17              S         05       10058.0   9010745.0   
...             ...            ...        ...           ...         ...   
538558      2018-19              S         54       71993.0    124776.0   
538759      2018-19              S         54       72249.0    107649.0   
540104      2018-19              S         54       76794.0   5430285.0   
540327      2018-19              S         55       10553.0    134569.0   
540355      2018-19              S         55       10553.0   5530118.0   
540374      2018-19              S         55       10553.0   5530142.0   
540396      2018-19              S         55       72389.0         0.0   
540579      2018-19              S         55       72413.0   5530050.0   
540589      2018-19              S         55       72413.0   5530092.0   
540605      2018-19              S         55       72413.0   5530100.0   
540621      2018-19              S         55       72413.0   5530175.0   
541019      2018-19              S         56       10561.0   6069231.0   
541182      2018-19              S         56       72520.0         1.0   
541204      2018-19              S         56       72520.0   5630405.0   
541322      2018-19              S         56       72546.0         1.0   
541332      2018-19              S         56       72546.0    115105.0   
541917      2018-19              S         56       72603.0         1.0   
542198      2018-19              S         56       72652.0         1.0   
542479      2018-19              S         56       73759.0         1.0   
542750      2018-19              S         56       73874.0         1.0   
543029      2018-19              S         56       76828.0         0.0   
543188      2018-19             

In [31]:
#Drop SuppressedTA
GRminusSup = GRminusSur.drop(GRminusSur.loc[(GRminusSur['CohortStudents'] == "*") & (GRminusSur['ReportingCategory'] == "TA") 
                         & (GRminusSur['AggregateLevel'] == "S") & (GRminusSur['CharterSchool'] != "All")
                            & (GRminusSur['DASS'] != "All")].index).reset_index(drop=True)

GRminusSup

AcademicYear AggregateLevel CountyCode  DistrictCode  SchoolCode  \
0           2016-17              C         01           NaN         NaN   
1           2016-17              C         01           NaN         NaN   
2           2016-17              C         01           NaN         NaN   
3           2016-17              C         01           NaN         NaN   
4           2016-17              C         01           NaN         NaN   
5           2016-17              C         01           NaN         NaN   
6           2016-17              C         01           NaN         NaN   
7           2016-17              C         01           NaN         NaN   
8           2016-17              C         01           NaN         NaN   
9           2016-17              C         01           NaN         NaN   
10          2016-17              C         01           NaN         NaN   
11          2016-17              C         01           NaN         NaN   
12          2016-17              C         01           NaN         NaN   
13          2016-17              C         01           NaN         NaN   
14          2016-17              C         01           NaN         NaN   
15          2016-17              C         01           NaN         NaN   
16          2016-17              C         01           NaN         NaN   
17          2016-17              C         01           NaN         NaN   
18          2016-17              C         01           NaN         NaN   
19          2016-17              C         01           NaN         NaN   
20          2016-17              C         01           NaN         NaN   
21          2016-17              C         01           NaN         NaN   
22          2016-17              C         01           NaN         NaN   
23          2016-17              C         01           NaN         NaN   
24          2016-17              C         01           NaN         NaN   
25          2016-17              C         01           NaN         NaN   
26          2016-17              C         01           NaN         NaN   
27          2016-17              C         01           NaN         NaN   
28          2016-17              C         01           NaN         NaN   
29          2016-17              C         01           NaN         NaN   
...             ...            ...        ...           ...         ...   
543127      2018-19              T         00           NaN         NaN   
543128      2018-19              T         00           NaN         NaN   
543129      2018-19              T         00           NaN         NaN   
543130      2018-19              T         00           NaN         NaN   
543131      2018-19              T         00           NaN         NaN   
543132      2018-19              T         00           NaN         NaN   
543133      2018-19              T         00           NaN         NaN   
543134      2018-19              T         00           NaN         NaN   
543135      2018-19              T         00           NaN         NaN   
543136      2018-19              T         00           NaN         NaN   
543137      2018-19              T         00           NaN         NaN   
543138      2018-19              T         00           NaN         NaN   
543139      2018-19              T         00           NaN         NaN   
543140      2018-19              T         00           NaN         NaN   
543141      2018-19              T         00           NaN         NaN   
543142      2018-19              T         00           NaN         NaN   
543143      2018-19              T         00           NaN         NaN   
543144      2018-19              T         00           NaN         NaN   
543145      2018-19              T         00           NaN         NaN   
543146      2018-19              T         00           NaN         NaN   
543147      2018-19              T         00           NaN         NaN   
543148      2018-19             

In [32]:
# GRminusSup.loc[(GRminusSup['CohortStudents'] == "*") & (GRminusSup['ReportingCategory'] == "TA") 
#                          & (GRminusSup['AggregateLevel'] == "S") & (GRminusSup['CharterSchool'] != "All")
#                             & (GRminusSup['DASS'] != "All")]

In [33]:
#Replace the wildcard * with null so dtype can be changed to numeric
Replacedstatic = GRminusSup.replace(to_replace = r"*", value=np.nan)

In [34]:
#Replacedstatic.loc[43]

In [35]:
GRminusSup.describe()

DistrictCode    SchoolCode
count  519404.000000  5.194040e+05
mean    62924.968770  1.525329e+06
std     16859.658939  1.853178e+06
min     10017.000000  0.000000e+00
25%     64287.000000  0.000000e+00
50%     67082.000000  1.339750e+05
75%     70524.000000  3.038023e+06
max     77222.000000  9.010745e+06

In [36]:
#COnvert other columns to Integer given they are calculated fields that we will need for our analysis
Replacedstatic[['CohortStudents', 'Regular HS Diploma Graduates (Count)', "Met UC/CSU Grad Req's (Count)", 'Seal of Biliteracy (Count)',
            'Golden State Seal Merit Diploma (Count)', 'CHSPE Completer (Count)', 'Adult Ed. HS Diploma (Count)', 'SPED Certificate (Count)',
            'GED Completer (Count)', 'Other Transfer (Count)', 'Dropout (Count)', 'Still Enrolled (Count)']] = Replacedstatic[['CohortStudents', 'Regular HS Diploma Graduates (Count)', "Met UC/CSU Grad Req's (Count)", 'Seal of Biliteracy (Count)',
            'Golden State Seal Merit Diploma (Count)', 'CHSPE Completer (Count)', 'Adult Ed. HS Diploma (Count)', 'SPED Certificate (Count)',
            'GED Completer (Count)', 'Other Transfer (Count)', 'Dropout (Count)', 'Still Enrolled (Count)']].apply(pd.to_numeric,  downcast = "signed")
print(Replacedstatic.dtypes)

AcademicYear                                object
AggregateLevel                              object
CountyCode                                  object
DistrictCode                               float64
SchoolCode                                 float64
CountyName                                  object
DistrictName                                object
SchoolName                                  object
CharterSchool                               object
DASS                                        object
ReportingCategory                           object
CohortStudents                             float64
Regular HS Diploma Graduates (Count)       float64
Met UC/CSU Grad Req's (Count)              float64
Seal of Biliteracy (Count)                 float64
Golden State Seal Merit Diploma (Count)    float64
CHSPE Completer (Count)                    float64
Adult Ed. HS Diploma (Count)               float64
SPED Certificate (Count)                   float64
GED Completer (Count)          

In [37]:
Replacedstatic.describe()

DistrictCode    SchoolCode  CohortStudents  \
count  519404.000000  5.194040e+05   321879.000000   
mean    62924.968770  1.525329e+06      295.961476   
std     16859.658939  1.853178e+06     4318.809387   
min     10017.000000  0.000000e+00       11.000000   
25%     64287.000000  0.000000e+00       25.000000   
50%     67082.000000  1.339750e+05       57.000000   
75%     70524.000000  3.038023e+06      168.000000   
max     77222.000000  9.010745e+06   504073.000000   

       Regular HS Diploma Graduates (Count)  Met UC/CSU Grad Req's (Count)  \
count                         321879.000000                  321879.000000   
mean                             242.332401                     113.656576   
std                             3726.790490                    1861.782374   
min                                0.000000                       0.000000   
25%                               16.000000                       0.000000   
50%                               40.000000                      10.000000   
75%                              132.000000                      56.000000   
max                           418205.000000                  210980.000000   

       Seal of Biliteracy (Count)  Golden State Seal Merit Diploma (Count)  \
count               321879.000000                            321879.000000   
mean                    25.780011                                51.620047   
std                    442.835576                               857.212442   
min                      0.000000                                 0.000000   
25%                      0.000000                                 0.000000   
50%                      0.000000                                 1.000000   
75%                      8.000000                                19.000000   
max                  48311.000000                            100922.000000   

       CHSPE Completer (Count)  Adult Ed. HS Diploma (Count)  \
count            321879.000000                 321879.000000   
mean                  0.753858                      0.285999   
std                  11.291795                      4.823179   
min                   0.000000                      0.000000   
25%                   0.000000                      0.000000   
50%                   0.000000                      0.000000   
75%                   0.000000                      0.000000   
max                1472.000000                    456.000000   

       SPED Certificate (Count)  GED Completer (Count)  \
count             321879.000000          321879.000000   
mean                   3.015571               0.308038   
std                   46.529071               4.427915   
min                    0.000000               0.000000   
25%                    0.000000               0.000000   
50%                    0.000000               0.000000   
75%                    1.000000               0.000000   
max                 4149.000000             564.000000   

       Other Transfer (Count)  Dropout (Count)  Still Enrolled (Count)  
count           321879.000000    321879.000000           321879.000000  
mean                 4.146437        28.639945               16.479227  
std                 56.602403       374.954729              211.631448  
min                  0.000000         0.000000                0.000000  
25%                  0.000000         1.000000                0.000000  
50%                  0.000000         5.000000                2.000000  
75%                  2.000000        13.000000                7.000000  
max               7356.000000     48453.000000            27023.000000

In [38]:
pd.set_option('precision', 0)
Replacedstatic.head(5).append(Replacedstatic.tail(5))

AcademicYear AggregateLevel CountyCode  DistrictCode  SchoolCode  \
0           2016-17              C         01           NaN         NaN   
1           2016-17              C         01           NaN         NaN   
2           2016-17              C         01           NaN         NaN   
3           2016-17              C         01           NaN         NaN   
4           2016-17              C         01           NaN         NaN   
543152      2018-19              T         00           NaN         NaN   
543153      2018-19              T         00           NaN         NaN   
543154      2018-19              T         00           NaN         NaN   
543155      2018-19              T         00           NaN         NaN   
543156      2018-19              T         00           NaN         NaN   

       CountyName DistrictName SchoolName CharterSchool DASS  \
0         Alameda          NaN        NaN           All  All   
1         Alameda          NaN        NaN           All  All   
2         Alameda          NaN        NaN           All  All   
3         Alameda          NaN        NaN           All  All   
4         Alameda          NaN        NaN           All  All   
543152      State          NaN        NaN           Yes  Yes   
543153      State          NaN        NaN           Yes  Yes   
543154      State          NaN        NaN           Yes  Yes   
543155      State          NaN        NaN           Yes  Yes   
543156      State          NaN        NaN           Yes  Yes   

       ReportingCategory  CohortStudents  \
0                     GF            8039   
1                     GM            8675   
2                     RA            3632   
3                     RB            2091   
4                     RD              62   
543152                SF            1070   
543153                SH            3627   
543154                SM             186   
543155                SS           21735   
543156                TA           26023   

        Regular HS Diploma Graduates (Count)  Met UC/CSU Grad Req's (Count)  \
0                                       7114                           4439   
1                                       7175                           3734   
2                                       3376                           2586   
3                                       1563                            589   
4                                         35                             20   
543152                                   267                             10   
543153                                   782                             52   
543154                                    63                              1   
543155                                  5643                            552   
543156                                  6605                            705   

        Seal of Biliteracy (Count)  Golden State Seal Merit Diploma (Count)  \
0                              989                                     1928   
1                              599                                     1567   
2                              685                                     1570   
3                               30                                      112   
4                                1                                        4   
543152                           0                                        2   
543153                           0                                       13   
543154                           0                                        0   
543155                          13                                      126   
543156                          14                                      197   

        CHSPE Completer (Count)  Adult Ed. HS Diploma (Count)  \
0                            30                             2   
1                            28                             2   
2                             9                         

In [39]:
Replacedstatic.columns

Index(['AcademicYear', 'AggregateLevel', 'CountyCode', 'DistrictCode',
       'SchoolCode', 'CountyName', 'DistrictName', 'SchoolName',
       'CharterSchool', 'DASS', 'ReportingCategory', 'CohortStudents',
       'Regular HS Diploma Graduates (Count)', 'Met UC/CSU Grad Req's (Count)',
       'Seal of Biliteracy (Count)', 'Golden State Seal Merit Diploma (Count)',
       'CHSPE Completer (Count)', 'Adult Ed. HS Diploma (Count)',
       'SPED Certificate (Count)', 'GED Completer (Count)',
       'Other Transfer (Count)', 'Dropout (Count)', 'Still Enrolled (Count)'],
      dtype='object')

In [40]:
#Rename select column names to make them more readable and error-prone
Replacedstatic.rename(columns={"Regular HS Diploma Graduates (Count)": "RegHSDiplomaGrad", "Met UC/CSU Grad Req's (Count)":"MetUCCSUGradReq",
                              "Seal of Biliteracy (Count)": "SealOfBiliteracy", "Golden State Seal Merit Diploma (Count)": "GSSealMeritDiploma",
                               "CHSPE Completer (Count)": "CHSPEComp", "Adult Ed. HS Diploma (Count)": "AdEdHSDiploma", "SPED Certificate (Count)":"SPEDCert",
                               "GED Completer (Count)": "GEDComp", "Other Transfer (Count)": "OtherTransfer", "Dropout (Count)": "Dropout",
                               "Still Enrolled (Count)": "StillEnrolled"}, inplace=True)
Replacedstatic.columns

Index(['AcademicYear', 'AggregateLevel', 'CountyCode', 'DistrictCode',
       'SchoolCode', 'CountyName', 'DistrictName', 'SchoolName',
       'CharterSchool', 'DASS', 'ReportingCategory', 'CohortStudents',
       'RegHSDiplomaGrad', 'MetUCCSUGradReq', 'SealOfBiliteracy',
       'GSSealMeritDiploma', 'CHSPEComp', 'AdEdHSDiploma', 'SPEDCert',
       'GEDComp', 'OtherTransfer', 'Dropout', 'StillEnrolled'],
      dtype='object')

In [41]:
#Change the datatype to ensure they are assigned dtypes that make their values useful

<font color = "blue"> As part of our next steps, </font>
* we need to further explore this dataset to ensure the relevance of all the fields we have. For example, do we want to keep the "DASS" column? If yes, we will need to further exolore what DASS means and it's specific relevance for our usecase. 
* Also, as noted above, submission has been cleaned with major emphasis on being able to report on school level data. The principles applied here can be reevaluated to extract for the other Category levels.
* The null values in District Code/District Name and School Code/School Name have been left as is given their missingness is intentional. We can explore whether to replace them with a specific value for the sake of keeping the technical principle of no missing data. Same applies for rethinking the * that were reset to null.
* Find common columns that can serve as basis for merging this data with the Master Expenses data and proceed with some deeper level of EDA

In [42]:
#A final glance at our cleaned data
Replacedstatic.info()
Replacedstatic.head(10)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 543157 entries, 0 to 543156
Data columns (total 23 columns):
AcademicYear          543157 non-null object
AggregateLevel        543157 non-null object
CountyCode            543157 non-null object
DistrictCode          519404 non-null float64
SchoolCode            519404 non-null float64
CountyName            543157 non-null object
DistrictName          519404 non-null object
SchoolName            519404 non-null object
CharterSchool         543157 non-null object
DASS                  543157 non-null object
ReportingCategory     543157 non-null object
CohortStudents        321879 non-null float64
RegHSDiplomaGrad      321879 non-null float64
MetUCCSUGradReq       321879 non-null float64
SealOfBiliteracy      321879 non-null float64
GSSealMeritDiploma    321879 non-null float64
CHSPEComp             321879 non-null float64
AdEdHSDiploma         321879 non-null float64
SPEDCert              321879 non-null float64
GEDComp               32

AcademicYear AggregateLevel CountyCode  DistrictCode  SchoolCode CountyName  \
0      2016-17              C         01           NaN         NaN    Alameda   
1      2016-17              C         01           NaN         NaN    Alameda   
2      2016-17              C         01           NaN         NaN    Alameda   
3      2016-17              C         01           NaN         NaN    Alameda   
4      2016-17              C         01           NaN         NaN    Alameda   
5      2016-17              C         01           NaN         NaN    Alameda   
6      2016-17              C         01           NaN         NaN    Alameda   
7      2016-17              C         01           NaN         NaN    Alameda   
8      2016-17              C         01           NaN         NaN    Alameda   
9      2016-17              C         01           NaN         NaN    Alameda   

  DistrictName SchoolName CharterSchool DASS ReportingCategory  \
0          NaN        NaN           All  All                GF   
1          NaN        NaN           All  All                GM   
2          NaN        NaN           All  All                RA   
3          NaN        NaN           All  All                RB   
4          NaN        NaN           All  All                RD   
5          NaN        NaN           All  All                RF   
6          NaN        NaN           All  All                RH   
7          NaN        NaN           All  All                RI   
8          NaN        NaN           All  All                RP   
9          NaN        NaN           All  All                RT   

   CohortStudents  RegHSDiplomaGrad  MetUCCSUGradReq  SealOfBiliteracy  \
0            8039              7114             4439               989   
1            8675              7175             3734               599   
2            3632              3376             2586               685   
3            2091              1563              589                30   
4              62                35               20                 1   
5             993               934              548                56   
6            5487              4396             1974               432   
7              45                33               14                 4   
8             212               170               59                 3   
9             643               585              356                60   

   GSSealMeritDiploma  CHSPEComp  AdEdHSDiploma  SPEDCert  GEDComp  \
0                1928         30              2        33        3   
1                1567         28              2        57        3   
2                1570          9              0        13        0   
3                 112          5              2        17        0   
4                   4          1              0         0        0   
5                 278          1              0         7        0   
6                 448          6              1        27        3   
7                   4          1              0         0        0   
8                  22          1              0         1        0   
9                 175          1              1         1        0   

   OtherTransfer  Dropout  StillEnrolled  
0             48      506            303  
1             70      836            504  
2             18      113            103  
3             19      291            194  
4              4       15              7  
5              3       29             19  
6             51      640            363  
7              3        5              3  
8              2       27             11  
9              1       36             18

In [43]:
combinedGR.to_csv('combinedGradRateData.csv', index=False)

In [44]:
Replacedstatic.to_csv('ReshapedGradRate.csv', index=False)